src/Language/PureScript/TypeChecker/Entailment/IntCompare.hs

```diff
@@ -0,0 +1,173 @@
+  -- and when verifying if a comparison holds true:
+  -- > solveComparison context (a .=. c)
+  --
+  -- the context and comparison is renamed as:
+  -- > [ abc .=. abc, xyz .=. xyz ] & (abc .=. abc)
+  Equal a b
+    | a == b -> True
+    | otherwise ->
+        let
+          comparison' = alpha comparison
+          context' = alpha <$> context
+        in
+          comparison' `elem` context'
+  -- Proving that an inequality exists between two nominal values takes
+  -- a very different code path compared to Equal; to summarize, it:
+  --
+  -- 1. Creates a directed graph of all inequalities in the renamed
+  -- context.
+  --
+  -- 2. Searches whether a path exists from point a to point b, allowing
+  -- symmetry to be solved.
+  --
+  -- 3. Also searches whether a path exists from point b to point a,
+  -- solving both the symmetry and reflexivity properties.
+  LessThan a b ->
+    solveInequality fst a b
+  GreaterThan a b ->
+    solveInequality snd a b
+  where
+  solveInequality
+    :: ( ( (G.Graph, S.Set a -> Maybe G.Vertex)
+         , (G.Graph, S.Set a -> Maybe G.Vertex)
+         ) -> (G.Graph, S.Set a -> Maybe G.Vertex)
+       )
+    -> a
+    -> a
+    -> Bool
+  solveInequality f a b =
+    let (graph, search) = f inequalities
+    in fromMaybe False $ do
+      a' <- search $ rename a
+      b' <- search $ rename b
+      let norm = G.path graph a' b'
+          symm = G.path (G.transposeG graph) b' a'
+      pure $ norm || symm
+
+  alpha :: Comparison a -> Comparison (S.Set a)
+  alpha = \case
+    Equal a b       -> Equal (rename a) (rename b)
+    LessThan a b    -> LessThan (rename a) (rename b)
+    GreaterThan a b -> GreaterThan (rename a) (rename b)
+
+  -- Determine which "equality bucket" a value lands on, then replace
+  -- that value using that bucket.
+  rename :: a -> S.Set a
+  rename n = fromMaybe (S.singleton n) (find (S.member n) equalities)
+
+  -- This determines which names are equal within the context, each set
+  -- in this list is a combination of all equal names, and it's built by
+  -- deliberately creating a cyclic graph between the equalities so as
+  -- to group them together.
+  equalities :: [S.Set a]
+  equalities = fmap (S.fromList . G.flattenSCC) $ makeStrong $ clean $ foldMap asNode context
+    where
+    asNode :: Comparison a -> [(a, [a])]
+    asNode = \case
+      Equal a b -> [(a, [b]), (b, [a])]
+      _         -> []
+
+    makeStrong :: [(a, [a])] -> [G.SCC a]
+    makeStrong = G.stronglyConnComp . fmap make
+
+  -- This determines the directed graphs for the "less than" and
+  -- "greater than" inequalities. This is fairly terse but what
+  -- it essentially does is:
+  --
+  -- 1. Collects all inequalities, renaming them on the way.
+  --
+  -- 2. Creates graphs for less than and greater than inequalities.
+  inequalities ::
+    ( ( G.Graph, S.Set a -> Maybe G.Vertex )
+    , ( G.Graph, S.Set a -> Maybe G.Vertex )
+    )
+  inequalities = lts &&& gts $ foldMap convert context
```

Comment:
I think you've overcomplicated this. There only needs to be one inequality graph, with edges pointing from lesser to greater, and you can look for a path from `a` to `b` when solving `LessThan` and from `b` to `a` when solving `GreaterThan` (if we even keep the `GreaterThan` constructor). That should let you simplify things here quite a bit, I think.

Reply:
I think so too, I can also add failing tests to be sure.